constantine/math_compiler/pub_fields.nim

```diff
@@ -24,41 +210,879 @@ proc genFpAdd*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## with signature
+  ##   void name(*bool r, FieldType a)
+  ## with r the result and a the operand
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = fd.name & "_isOdd_internal"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r, a]),
+          {kHot}):
+    tagParameter(1, "sret")
+
+    let (ri, ai) = llvmParams
+    let aA = asy.asArray(ai, fd.fieldTy)
+
+    # Check if the least significant bit of the first word is 1
+    let lsb = asy.br.and(aA[0], constInt(fd.wordTy, 1))
+    let isOddI1 = asy.br.icmp(kNE, lsb, constInt(fd.wordTy, 0))
+
+    asy.store(ri, isOddI1)
+    asy.br.retVoid()
+
+  asy.callFn(name, [r, a])
+
+proc genFpIsOdd*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## Generate a public field isOdd proc
+  ## with signature
+  ##   void name(*bool r, FieldType a)
+  ## with r the result and a the operand
+  ## and return the corresponding name to call it
+
+  let name = fd.name & "_isOdd"
+  let ptrBool = pointer_t(asy.ctx.int1_t())
+  asy.llvmPublicFnDef(name, "ctt." & fd.name, asy.void_t, [ptrBool, fd.fieldTy]):
+    let (r, a) = llvmParams
+    asy.isOdd_internal(fd, r, a)
+    asy.br.retVoid()
+
+  return name
+
+proc neg_internal*(asy: Assembler_LLVM, fd: FieldDescriptor, r, a: ValueRef) {.used.} =
+  ## Generate an internal field negation
+  ## with signature
+  ##   `void name(FieldType r, FieldType a)`
+  ## with `r` the resulting negated field element.
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = fd.name & "_neg_internal"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r, a]),
+          {kHot}):
+
+    tagParameter(1, "sret")
+    let (ri, ai) = llvmParams
+    let M = asy.getModulusPtr(fd)
+    let aA = asy.asArray(ai, fd.fieldTy)
+    let rA = asy.asArray(ri, fd.fieldTy)
+
+    # Subtraction M - a
+    asy.modsub(fd, ri, M, ai, M)
+
+    # Determine if `a == 0`
+    var isZero = aA[0]
+    for i in 1 ..< fd.numWords:
+      isZero = asy.br.`or`(isZero, aA[i])
+
+    # `slct` takes a condition and returns argument 1 if `cond >= 0` and argument 2 if `cond < 0`.
+    # Given that our `isZero` value is either positive or 0, we need to split the `== 0` and `> 0`
+    # for the condition. We do this by turning positive values into negative ones and keeping
+    # zeroes as is.
+    # Further, `slct` needs a 32 bit value as the condition. We want to avoid truncating `isZero`,
+    # from potentially larger values.
+    # So we create a boolean of whether it is 0, extend it to 32 bit and
+    # compute `0 - X` with `X ∈ {0, 1}`.
+
+    # create an int1 bool indicating if isZero is not zero: `notZero = isZero != 0`
+    let cZeroW = constInt(fd.wordTy, 0)
+    let notZero = asy.br.icmp(kNE, isZero, cZeroW)
+
+    # extend `int1_t` to `int32_t`
+    let notZeroExt = asy.br.zext(notZero, asy.ctx.int32_t())
```

Comment:
if we have to do this, we can just use the LLVM builtin select instead. Then the compiled should turn it in either conditional mov or predicated mov (or branches but not on a GPU).